knife/lib/chef/knife/ssh.rb

```diff
@@ -358,21 +364,22 @@ def ssh_command(command, subsession = nil)
         subsession ||= session
         command = fixup_sudo(command)
         command.force_encoding("binary") if command.respond_to?(:force_encoding)
-        begin
-          open_session(subsession, command)
-        rescue => e
-          open_session(subsession, command, true)
-        end
+        open_session(subsession, command)
       end
 
-      def open_session(subsession, command, pty = false)
+      def open_session(subsession, command)
```

Comment:
This needs to be renamed to something like `open_next_sesssion`.

The root of the problem here is that `subsession ||= session` is set in the caller so the "subsession" is actually the whole session container and calling "open_channel" on that gets the next one.  So the rescue was always proceeding with the next host in the list and never retrying (which is the "race condition").

Reply:
`subsession` is also really `session_list` or something like that (where a single subsession is a list of 1 the way the net-ssh library uses it I guess)